File: scripts/wave1-banks/write-helper.mjs
const ANSWERS = ["a", "b", "c", "d"];
export const SOURCE_NOTE = "Original UniPrep2Go local bank (Wave 1).";

/** @param {number} i 0-based index */
export function rot(i) {
  return ANSWERS[i % 4];
}

/**
 * @param {string} examSlug
 * @param {string} topicId
 * @param {number} n 1-based
 * @param {[string, string, string[], string, string[], string]} row
 *   [prompt, correct, wrongs[3], explanation, distractorNotes[3], difficulty]
 * @param {string} [correctId]
 */
export function makeQ(examSlug, topicId, n, row, correctId = "a") {
  const [prompt, correct, wrongs, explanation, distNotes, difficulty] = row;
  if (wrongs.length !== 3 || distNotes.length !== 3) {
    throw new Error(`Bad row ${examSlug} ${topicId} ${n}`);
  }
  const wrongIds = ANSWERS.filter((id) => id !== correctId);
  const options = ANSWERS.map((id) => ({
    id,
    text: id === correctId ? correct : wrongs[wrongIds.indexOf(id)],
  }));
  const distractorExplanations = {};
  wrongIds.forEach((id, i) => {
    distractorExplanations[id] = distNotes[i];
  });
  const num = String(n).padStart(3, "0");
  return {
    id: `${examSlug}-${topicId}-${num}`,
    examSlug,
    topicId,
    prompt,
    options,
    correctOptionId: correctId,
    explanation,
    distractorExplanations,
    difficulty,
    sourceNote: SOURCE_NOTE,
  };
}

/** Wave 1 session banks: 4 topics × 10 questions = 40. */
export function buildBank(examSlug, topics, perTopic = 10) {
  /** @type {any[]} */
  const out = [];
  const topicIds = Object.keys(topics);
  if (topicIds.length === 0) throw new Error(`${examSlug}: no topics`);
  for (const topicId of topicIds) {
    const rows = topics[topicId];
    if (rows.length !== perTopic) {
      throw new Error(`${examSlug}/${topicId}: expected ${perTopic}, got ${rows.length}`);
    }
    rows.forEach((row, i) => {
      out.push(makeQ(examSlug, topicId, i + 1, row, rot(i)));
    });
  }
  const expectedTotal = perTopic * topicIds.length;
  if (out.length !== expectedTotal) {
    throw new Error(`${examSlug}: expected ${expectedTotal}, got ${out.length}`);
  }
  return out;
}

File: scripts/wave2-banks/write_helper.py
#!/usr/bin/env python3
"""Wave 2 mock-bank writer: 4 topics × 15 Q = 60."""

from __future__ import annotations

import json
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
OUT = ROOT / "src/data/mock-exams"
SOURCE = "Original UniPrep2Go local bank (Wave 2)."
ANSWERS = ("a", "b", "c", "d")


def make_q(exam_slug: str, topic_id: str, n: int, row: tuple, correct_id: str) -> dict:
    prompt, correct, wrongs, explanation, dist_notes, difficulty = row
    if len(wrongs) != 3 or len(dist_notes) != 3:
        raise SystemExit(f"Bad row {exam_slug}/{topic_id}/{n}")
    wrong_ids = [i for i in ANSWERS if i != correct_id]
    options = []
    for oid in ANSWERS:
        if oid == correct_id:
            options.append({"id": oid, "text": correct})
        else:
            options.append({"id": oid, "text": wrongs[wrong_ids.index(oid)]})
    distractors = {wid: dist_notes[i] for i, wid in enumerate(wrong_ids)}
    return {
        "id": f"{exam_slug}-{topic_id}-{n:03d}",
        "examSlug": exam_slug,
        "topicId": topic_id,
        "prompt": prompt,
        "options": options,
        "correctOptionId": correct_id,
        "explanation": explanation,
        "distractorExplanations": distractors,
        "difficulty": difficulty,
        "sourceNote": SOURCE,
    }


def build_bank(exam_slug: str, topics: dict[str, list]) -> list[dict]:
    out: list[dict] = []
    if len(topics) != 4:
        raise SystemExit(f"{exam_slug}: expected 4 topics, got {len(topics)}")
    for topic_id, rows in topics.items():
        if len(rows) != 15:
            raise SystemExit(f"{exam_slug}/{topic_id}: expected 15, got {len(rows)}")
        for i, row in enumerate(rows):
            out.append(make_q(exam_slug, topic_id, i + 1, row, ANSWERS[i % 4]))
    if len(out) != 60:
        raise SystemExit(f"{exam_slug}: expected 60, got {len(out)}")
    return out


def write_bank(exam_slug: str, topics: dict[str, list]) -> None:
    bank = build_bank(exam_slug, topics)
    path = OUT / f"{exam_slug}.json"
    path.write_text(json.dumps(bank, indent=2) + "\n", encoding="utf-8")
    counts = {}
    for q in bank:
        counts[q["topicId"]] = counts.get(q["topicId"], 0) + 1
        wrong = [x for x in ANSWERS if x != q["correctOptionId"]]
        for wid in wrong:
            if wid not in q["distractorExplanations"]:
                raise SystemExit(f"Missing distractor {wid} on {q['id']}")
        if q["correctOptionId"] in q["distractorExplanations"]:
            raise SystemExit(f"Correct in distractors on {q['id']}")
        if q["sourceNote"] != SOURCE:
            raise SystemExit(f"Bad sourceNote on {q['id']}")
    print(f"{path.name}: {len(bank)}", counts)
